File: lester/examples/highlyrated_books/encode_with_sparkml.py
from lester.context import split, encode_features, DataframeDialect, EstimatorTransformerDialect

from pyspark.ml import Pipeline
from pyspark.ml.feature import HashingTF, Tokenizer, StringIndexer, OneHotEncoder, StandardScaler, VectorAssembler


@split(dialect=DataframeDialect.PYSPARK)
def random_split(data, random_seed):
    return data.randomSplit([0.8, 0.2], seed=random_seed)


@encode_features(dialect=EstimatorTransformerDialect.SPARKML)
def encode_books():
    stages = []

    # One-hot encode categorical features
    categorical_columns = ['authors', 'tag_id', 'original_publication_year']
    for column in categorical_columns:
        stages.append(StringIndexer(inputCol=column, outputCol=f"{column}Index", handleInvalid='keep'))

    categorical_indexes = [f'{column}Index' for column in categorical_columns]
    categorical_features = [f'{column}Vec' for column in categorical_columns]
    encoder = OneHotEncoder(inputCols=categorical_indexes, outputCols=categorical_features)
    stages.append(encoder)

    numerical_columns = ['work_text_reviews_count']

    numerical_assembler = VectorAssembler(inputCols=numerical_columns, outputCol='numerical_features_raw')
    stages.append(numerical_assembler)
    # Normalize numerical features
    scaler = StandardScaler(inputCol='numerical_features_raw', outputCol='numerical_features', withMean=True)
    stages.append(scaler)

    # Hash word features
    tokenizer = Tokenizer(inputCol="title", outputCol="words")
    hashing = HashingTF(inputCol="words", numFeatures=100, outputCol="text_features")

    stages.append(tokenizer)
    stages.append(hashing)

    # Concatenate all features
    assembler = VectorAssembler(inputCols=categorical_features + ['numerical_features', 'text_features'],
                                outputCol="features")
    stages.append(assembler)

    return Pipeline(stages=stages)

File: src/main/scala/berlin/deem/lester/example/highlyratedbooks/EncodeWithSparkML.scala
package berlin.deem.lester.example.highlyratedbooks

import berlin.deem.lester.context.{EncodeFeatures, Split}

import scala.collection.mutable.ArrayBuffer
import org.apache.spark.ml.{Pipeline, PipelineStage}
import org.apache.spark.ml.feature.{HashingTF, OneHotEncoder, StandardScaler, StringIndexer, Tokenizer, VectorAssembler}
import org.apache.spark.sql.DataFrame

class EncodeWithSparkML {

  @Split
  def trainTestSplit(data: DataFrame, randomSeed: Long): (DataFrame, DataFrame) = {
    val splits = data.randomSplit(Array(0.8, 0.2), randomSeed)
    val train = splits(0)
    val test = splits(1)

    (train, test)
  }

  @EncodeFeatures
  def encodeBooks(): Pipeline = {

    val stages = ArrayBuffer[PipelineStage]()

    val categoricalColumns = Array("tag_id", "original_publication_year")
    for (column <- categoricalColumns) {
      val indexer = new StringIndexer()
        .setInputCol(column)
        .setOutputCol(column + "Index")
        .setHandleInvalid("keep")

      stages += indexer
    }

    val categoricalIndexes = categoricalColumns.map(column => column + "Index")
    val categoricalFeatures = categoricalColumns.map(column => column + "Vec")

    val encoder = new OneHotEncoder()
      .setInputCols(categoricalIndexes)
      .setOutputCols(categoricalFeatures)

    stages += encoder

    val numericalColumns = Array("work_text_reviews_count")

    stages += new VectorAssembler()
      .setInputCols(numericalColumns)
      .setOutputCol("numericalFeaturesRaw")

    stages += new StandardScaler()
      .setInputCol("numericalFeaturesRaw")
      .setOutputCol("numericalFeatures")
      .setWithMean(true)

    stages += new Tokenizer()
      .setInputCol("title")
      .setOutputCol("words")

    stages += new HashingTF()
      .setInputCol("words")
      .setOutputCol("textFeatures")
      .setNumFeatures(100)

    stages += new VectorAssembler()
      .setInputCols(categoricalFeatures ++ Array("numericalFeatures", "textFeatures"))
      .setOutputCol("features")

    val featureTransform = new Pipeline()
      .setStages(stages.toArray)

    featureTransform
  }
}
